cidlists/cidlists.go

```diff
@@ -0,0 +1,111 @@
+package cidlists
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	datatransfer "github.com/filecoin-project/go-data-transfer"
+)
+
+// CIDLists maintains files that contain a list of CIDs received for different data transfers
+type CIDLists interface {
+	CreateList(chid datatransfer.ChannelID, initalCids []cid.Cid) error
+	AppendList(chid datatransfer.ChannelID, c cid.Cid) error
+	ReadList(chid datatransfer.ChannelID) ([]cid.Cid, error)
+	DeleteList(chid datatransfer.ChannelID) error
+}
+
+type cidLists struct {
+	baseDir string
+}
+
+// NewCIDLists initializes a new set of cid lists in a given directory
+func NewCIDLists(baseDir string) (CIDLists, error) {
+	base := filepath.Clean(string(baseDir))
+	info, err := os.Stat(string(base))
+	if err != nil {
+		return nil, fmt.Errorf("error getting %s info: %s", base, err.Error())
+	}
+	if !info.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", base)
+	}
+	return &cidLists{
+		baseDir: base,
+	}, nil
+}
+
+// CreateList initializes a new CID list with the given initial cids (or can be empty) for a data transfer channel
+func (cl *cidLists) CreateList(chid datatransfer.ChannelID, initialCids []cid.Cid) (err error) {
+	f, err := os.Create(transferFilename(cl.baseDir, chid))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		closeErr := f.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+	for _, c := range initialCids {
+		err := cbg.WriteCid(f, c)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AppendList appends a single CID to the list for a given data transfer channel
+func (cl *cidLists) AppendList(chid datatransfer.ChannelID, c cid.Cid) (err error) {
```

Comment:
Isn't this going to be really slow if we have to do this per cid? I'd suggest having some form of batching.

Reply:
Most files will be big, and the default lotus chunking is 1MB, so this should be usable

But it's still pretty far from ideal